File: sdk/typescript/rooch-sdk/src/utils/hash.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0
import { blake2b } from '@noble/hashes/blake2b'
import { ripemd160 } from '@noble/hashes/ripemd160'
import { sha3_256 } from '@noble/hashes/sha3'
import { sha256 } from '@noble/hashes/sha256'
import { sha512 } from '@noble/hashes/sha512'
import { utils as packedUtils } from 'micro-packed'

import { Bytes } from '@/types'

const { concatBytes } = packedUtils

export { sha256 }
export { sha3_256 }
export { sha512 }
export { blake2b }

export const hash160 = (msg: Bytes) => ripemd160(sha256(msg))
export const sha256x2 = (...msgs: Bytes[]) => sha256(sha256(concatBytes(...msgs)))

File: sdk/golang/rooch-sdk/utils/hash.go
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

package utils

import (
	"crypto/sha256"
	"golang.org/x/crypto/blake2b"
	"golang.org/x/crypto/ripemd160"
	"golang.org/x/crypto/sha3"
)

// Sha256 computes SHA256 hash of the input bytes
func Sha256(msg []byte) []byte {
	hash := sha256.New()
	hash.Write(msg)
	return hash.Sum(nil)
}

// Sha3_256 computes SHA3-256 hash of the input bytes
func Sha3_256(msg []byte) []byte {
	hash := sha3.New256()
	hash.Write(msg)
	return hash.Sum(nil)
}

// Sha512 computes SHA512 hash of the input bytes
func Sha512(msg []byte) []byte {
	hash := sha512.New()
	hash.Write(msg)
	return hash.Sum(nil)
}

// Blake2b computes BLAKE2b hash of the input bytes
func Blake2b(msg []byte) []byte {
	hash := blake2b.New512(nil)
	hash.Write(msg)
	return hash.Sum(nil)
}

// Hash160 computes RIPEMD160(SHA256(msg))
func Hash160(msg []byte) []byte {
	sha256Hash := Sha256(msg)
	ripemd := ripemd160.New()
	ripemd.Write(sha256Hash)
	return ripemd.Sum(nil)
}

// Sha256x2 computes SHA256(SHA256(concatenated msgs))
func Sha256x2(msgs ...[]byte) []byte {
	concatenated := ConcatBytes(msgs...)
	firstHash := Sha256(concatenated)
	return Sha256(firstHash)
} 